conductor_api/src/conductor/admin.rs

```diff
@@ -381,22 +388,70 @@ impl ConductorAdmin for Conductor {
         Ok(())
     }
 
-    fn add_agent(&mut self, new_agent: AgentConfiguration) -> Result<(), HolochainError> {
+    fn add_agent(
+        &mut self,
+        id: String,
+        name: String,
+        holo_remote_key: Option<bool>,
+    ) -> Result<String, HolochainError> {
         let mut new_config = self.config.clone();
-        if new_config.agents.iter().any(|i| i.id == new_agent.id) {
+        if new_config.agents.iter().any(|i| i.id == id) {
             return Err(HolochainError::ErrorGeneric(format!(
                 "Agent with ID '{}' already exists",
-                new_agent.id
+                id
             )));
         }
-        new_config.agents.push(new_agent.clone());
+
+        let (keystore, public_address) = if self.using_dpki() {
+            let dpki_instance_id = self.dpki_instance_id().unwrap();
+
+            // try to create the keystore first so that if the passphrase fails we don't have
+            // to clean-up any dkpi calls
+            let mut keystore =
+                Keystore::new(self.passphrase_manager.clone(), self.hash_config.clone())?;
+            {
+                let instance = self.instances.get(&dpki_instance_id)?;
+                let hc_lock = instance.clone();
+                let hc_lock_inner = hc_lock.clone();
+                let mut hc = hc_lock_inner.write().unwrap();
+                hc.dpki_create_agent_key(name.clone())?;
+            }
+            // TODO: how do we clean-up now if this fails? i.e. the dpki dna will have registered
+            // the identity to its DHT, but we failed, for what ever reason, to set up
+            // the agent in the conductor, so we should do something...
```

Comment:
Maybe we should not exit this function if we fail to access the DPKI keystore and regard it as a transient problem - and instead continue with adding the agent to the conductor config?